common/cluster/dynamic_metadata.go

```diff
@@ -0,0 +1,502 @@
+		config.MasterClusterName,
+		config.CurrentClusterName,
+		config.ClusterInformation,
+		clusterMetadataStore,
+		logger,
+	)
+}
+
+func NewDynamicMetadataForTest(
+	config *Config,
+) DynamicMetadata {
+	return NewDynamicMetadata(
+		config.EnableGlobalNamespace,
+		config.FailoverVersionIncrement,
+		config.MasterClusterName,
+		config.CurrentClusterName,
+		config.ClusterInformation,
+		nil,
+		log.NewNoopLogger(),
+	)
+}
+
+func (d *dynamicMetadataImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&d.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	err := d.refreshClusterMetadata(context.Background())
+	if err != nil {
+		d.logger.Fatal("Unable to initialize cluster metadata cache", tag.Error(err))
+	}
+	d.refresher = goro.Go(context.Background(), d.refreshLoop)
+}
+
+func (d *dynamicMetadataImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&d.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	d.refresher.Cancel()
+	<-d.refresher.Done()
+}
+
+func (d *dynamicMetadataImpl) IsGlobalNamespaceEnabled() bool {
+	return d.enableGlobalNamespace
+}
+
+func (d *dynamicMetadataImpl) IsMasterCluster() bool {
+	return d.masterClusterName == d.currentClusterName
+}
+
+func (d *dynamicMetadataImpl) GetNextFailoverVersion(clusterName string, currentFailoverVersion int64) int64 {
+	d.clusterLock.RLock()
+	defer d.clusterLock.RUnlock()
+
+	info, ok := d.clusterInfo[clusterName]
+	if !ok {
+		panic(fmt.Sprintf(
+			"Unknown cluster name: %v with given cluster initial failover version map: %v.",
+			clusterName,
+			d.clusterInfo,
+		))
+	}
+	failoverVersion := currentFailoverVersion/d.failoverVersionIncrement*d.failoverVersionIncrement + info.InitialFailoverVersion
+	if failoverVersion < currentFailoverVersion {
+		return failoverVersion + d.failoverVersionIncrement
+	}
+	return failoverVersion
+}
+
+func (d *dynamicMetadataImpl) IsVersionFromSameCluster(version1 int64, version2 int64) bool {
+	return (version1-version2)%d.failoverVersionIncrement == 0
+}
+
+func (d *dynamicMetadataImpl) GetMasterClusterName() string {
+	return d.masterClusterName
+}
+
+func (d *dynamicMetadataImpl) GetCurrentClusterName() string {
+	return d.currentClusterName
+}
+
+func (d *dynamicMetadataImpl) GetAllClusterInfo() map[string]ClusterInformation {
+	d.clusterLock.RLock()
+	defer d.clusterLock.RUnlock()
+
+	result := make(map[string]ClusterInformation, len(d.clusterInfo))
+	for k, v := range d.clusterInfo {
+		result[k] = v
+	}
+	return result
+}
+
+func (d *dynamicMetadataImpl) ClusterNameForFailoverVersion(failoverVersion int64) string {
+	if failoverVersion == common.EmptyVersion {
+		return d.currentClusterName
+	}
+
+	initialFailoverVersion := failoverVersion % d.failoverVersionIncrement
+	// Failover version starts with 1.  Zero is an invalid value for failover version
+	if initialFailoverVersion == common.EmptyVersion {
+		initialFailoverVersion = d.failoverVersionIncrement
+	}
```

Comment:
why not simply panic?

Reply:
This is the original logic. I guess we allow 0 failover version in some previous release and to make it compatible?